sogs/model/manager.py

```diff
@@ -0,0 +1,300 @@
+from __future__ import annotations
+
+from .. import crypto, db, config
+from ..db import query
+from ..web import app
+from .exc import BadPermission, PostRateLimited
+from .. import utils
+from .user import User
+from ..omq import omq_global
+from .bot import Bot
+from .room import Room
+from .message import Message
+from .filter import SimpleFilter
+from .exc import InvalidData
+import heapq
+
+from dataclasses import dataclass, field
+from typing import Optional, List, Union, Any
+import time
+
+"""
+Complications:
+    - given captcha bot
+        - new user (not approved) posts message
+        - we need bot to reply with whisper to that user with simple problem
+        - what does the bot do with the message they tried to send?
+            - can store locally
+            - user sends reply
+            - bot inserts it into room (?)
+
+Control Flow:
+    1) message comes in HTTP request
+    2) unpacked/parsed/verified/permissions checked
+    3) comes into relevant route (ex: add_post())
+    4) sends off to mule to be handled by bots
+    5) mule has ordered list of bots by priority
+    6) mule passes message to bots, which have fixed return values (insert, do not insert)
+    7) if all bots approve, mule replies to worker with go ahead or vice versa for no go
+"""
+
+
+@dataclass(order=True)
+class PriorityTuple(tuple):
+    priority: int
+    item: Any = field(compare=False)
+
+
+# Simple "priority queue" of bots implemented using a dict with heap
+# invariance maintained by qheap algorithm
+# TODO: when bots are designed basically, add methods for polling them
+#   and receiving their judgments
+class BotQueue:
+    def __init__(self) -> None:
+        self.queue = {}
+
+    def _qsize(self) -> int:
+        return len(self.queue.keys())
+
+    def _empty(self) -> bool:
+        return not self._qsize()
+
+    def _peek(self, priority: int):
+        return self.queue.get(priority)
+
+    def _put(self, item: PriorityTuple):
+        temp = list(self.queue.items())
+        heapq.heappush(temp, item)
+        self.queue = dict(temp)
+
+    def _get(self):
+        return heapq.heappop(self.queue)
+
+
+class Manager:
+    """
+    Class representing an interface that manages active bots
+
+    Object Properties:
+        queue - BotQueue object
+    """
+
+    def __init__(
+        self,
+        _rooms: List[Room],
+        row=None,
+        *,
+        id: Optional[int] = None,
+        session_id: Optional[int] = None,
+    ) -> None:
+        self.id = id
+        self.queue = BotQueue()
+
+    def qempty(self):
+        return not self.queue._empty()
+
+    def add_bot(self, bot: Bot, priority: int = None):
```

Comment:
A `Bot` here would be a connection to a bot rather than a Class that can be executed.